src/config_generator/from_proto.rs

```diff
@@ -0,0 +1,394 @@
+            if let Some(vars) = ty.variants {
+                variants.extend(vars);
+            }
+            ty.variants = Some(variants);
+            self.ty.types.insert(self.get(enum_name).unwrap(), ty);
+            // it should be
+            // safe to call
+            // unwrap here
+        }
+        self
+    }
+
+    /// Processes proto message types.
+    fn append_msg_type(mut self, messages: Vec<DescriptorProto>) -> Context<Config> {
+        if messages.is_empty() {
+            return self;
+        }
+        for message in messages {
+            let msg_name = message.name().to_string();
+
+            let mut ty = self.get_ty(&msg_name, DescriptorType::Message);
+
+            self = self.append_enums(message.enum_type);
+            self = self.append_msg_type(message.nested_type);
+
+            for field in message.field {
+                let field_name = field.name().to_string();
+                let mut cfg_field = Field::default();
+
+                let label = field.label().as_str_name().to_lowercase();
+                cfg_field.list = label.contains("repeated");
+                cfg_field.required = label.contains("required");
+
+                if field.r#type.is_some() {
+                    let type_of = convert_ty(field.r#type().as_str_name());
+                    cfg_field.type_of = type_of.to_string();
+                } else {
+                    // for non-primitive types
+                    let type_of = convert_ty(field.type_name());
+                    cfg_field.type_of = self.get(&type_of).unwrap_or(type_of);
+                }
+
+                ty.fields.insert(field_name, cfg_field);
+            }
+
+            self.ty.types.insert(self.get(&msg_name).unwrap(), ty); // it should
+                                                                    // be
+                                                                    // safe to call
+                                                                    // unwrap here
+        }
+        self
+    }
+
+    /// Generates argument configurations for service methods.
+    fn get_arg(&self, input_ty: &str) -> Option<(String, Arg)> {
+        match input_ty {
+            "google.protobuf.Empty" | "" => None,
+            any => {
+                let key = convert_ty(any).to_case(Case::Camel);
+                let val = Arg {
+                    type_of: self.get(any).unwrap_or(any.to_string()),
+                    list: false,
+                    required: true,
+                    /* Setting it not null by default. There's no way to infer this
+                     * from proto file */
+                    doc: None,
+                    modify: None,
+                    default_value: None,
+                };
+
+                Some((key, val))
+            }
+        }
+    }
+
+    /// Generates a Type configuration for service methods.
+    fn generate_ty(&mut self, services: Vec<ServiceDescriptorProto>, key: &str) -> Type {
```

Comment:
```suggestion
    fn generate_ty(mut self, services: Vec<ServiceDescriptorProto>, key: &str) -> Type {
```

No mut references.

Reply:
The design of this function should be — Either get the type or update self with the generated types. It's a little confusing what it does right now. If it generates stuff, then it should take in a `&self` if it updates then it should take `self`

Reply:
I think we don't need that function anymore.. earlier we decided to seperate query and mutation, both had almost the same code so I created this method but now it's not needed